File: Game/Scripts/Source/NodeFunctions/ScreenEffects.lua
local GlobalCore = require("GlobalCore")

local NativeScreenEffects = GlobalCore.ScreenEffects
local ScreenEffects = {}

function ScreenEffects.FlashScreen(red, green, blue, alpha, duration)
    red = red == nil and 255 or red
    green = green == nil and 255 or green
    blue = blue == nil and 255 or blue
    alpha = alpha == nil and 255 or alpha
    duration = duration == nil and 0.5 or duration
    NativeScreenEffects.flashScreen(sf.Color.new(red, green, blue, alpha), duration)
end

function ScreenEffects.StopFlashScreen()
    NativeScreenEffects.stopFlash()
end

function ScreenEffects.ChangeScreenTone(red, green, blue, gray, duration)
    red = red == nil and 0 or red
    green = green == nil and 0 or green
    blue = blue == nil and 0 or blue
    gray = gray == nil and 0 or gray
    duration = duration == nil and 0.0 or duration
    NativeScreenEffects.changeScreenTone(red, green, blue, gray, duration)
end

function ScreenEffects.ClearScreenTone(duration)
    duration = duration == nil and 0.0 or duration
    NativeScreenEffects.clearScreenTone(duration)
end

function ScreenEffects.ScreenShake(power, speed, duration)
    power = power == nil and 4.0 or power
    speed = speed == nil and 10.0 or speed
    duration = duration == nil and 0.5 or duration
    NativeScreenEffects.startShake(power, speed, duration)
end

function ScreenEffects.StopScreenShake()
    NativeScreenEffects.stopShake()
end

return ScreenEffects

File: Game/Engine/Source/Global/src/ScreenEffects.cpp
#include <ScreenEffects.hpp>
#include "System/FramePipelineImpl.hpp"
#include <utility>

void ScreenEffects::applyScreenTonePass() {
    ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .applyScreenTonePass(
            ludork::global::system_impl::framePipelineImpl().getCanvas());
}

void ScreenEffects::flashScreen(std::optional<sf::Color> color,
                                float duration) {
    ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .flashScreen(std::move(color), duration);
}

void ScreenEffects::stopFlash() {
    ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .stopFlash();
}

bool ScreenEffects::isFlashing() {
    return ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .isFlashing();
}

void ScreenEffects::changeScreenTone(float red, float green, float blue,
                                     float gray, float duration) {
    ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .changeScreenTone(red, green, blue, gray, duration);
}

void ScreenEffects::clearScreenTone(float duration) {
    ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .clearScreenTone(duration);
}

void ScreenEffects::stopScreenTone() {
    ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .stopScreenTone();
}

bool ScreenEffects::isScreenToneActive() {
    return ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .isScreenToneActive();
}

bool ScreenEffects::isScreenToneTransitionComplete() {
    return ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .isScreenToneTransitionComplete();
}

void ScreenEffects::startShake(float power, float speed, float duration) {
    ludork::global::system_impl::framePipelineImpl().screenEffects().startShake(
        power, speed, duration);
}

void ScreenEffects::stopShake() {
    ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .stopShake();
}

bool ScreenEffects::isShaking() {
    return ludork::global::system_impl::framePipelineImpl()
        .screenEffects()
        .isShaking();
}
